nobrainer/brainsiam_train.py

```diff
@@ -0,0 +1,47 @@
+"""
+Extending the SimSiam network architecture to brain volumes
+author: Dhritiman Das
+"""
+
+import tensorflow as tf
+
+
+class BrainSiamese(tf.keras.Model):
+    def __init__(self, encoder, predictor):
+        super(BrainSiamese, self).__init__()
+        self.encoder = encoder
+        self.predictor = predictor
+        self.loss_tracker = tf.keras.metrics.Mean(name="loss")
+
+    @property
+    def metrics(self):
+        return [self.loss_tracker]
+
+    def train_step(self, data):
+        data_one, data_two = data  # unpacking the data
+
+        # Forward pass through the encoder and predictor.
+        with tf.GradientTape() as tape:
+            proj_1, proj_2 = self.encoder(data_one), self.encoder(data_two)
+            pred_1, pred_2 = self.predictor(proj_1), self.predictor(proj_2)
+            loss = compute_loss(pred_1, proj_2) / 2 + compute_loss(pred_2, proj_1) / 2
+
+        # Compute gradients and update the parameters.
+        learnable_params = (
+            self.encoder.trainable_variables + self.predictor.trainable_variables
+        )
+        gradients = tape.gradient(loss, learnable_params)
+        self.optimizer.apply_gradients(zip(gradients, learnable_params))
+
+        # Monitor loss.
+        self.loss_tracker.update_state(loss)
+        return {"loss": self.loss_tracker.result()}
+
+
+def compute_loss(pred, proj):
```

Comment:
this seems to generic a name for a standalone function. unless you want others to use this function outside of simsiam, you can make this somewhat private by prepending with an underscore. i would also encourage changing the name.

Reply:
this is to compute the negative cosine similarity loss. I have modified the function name to make it more evident.